src/confluence/core/validate.py

```diff
@@ -0,0 +1,324 @@
+    and throws an error if one is found.
+    """
+    def __init__(self):
+        self._callbacks = []
+
+    def __call__(self, df):
+        """
+        Function: Cycles through each callback, feeding the
+        dataframe to the next function to check for errors.
+        :param df: dataframe to be tested
+        :param filename: Filename corresponding to the dataframe
+        :param sheet: Sheet corresponding to the dataframe
+        :return: dataframe with no errors
+        """
+        for callback in self._callbacks:
+            df = callback(df)
+        return df
+
+    def add_callback(self, func):
+        # if not hasattr('__call__', func):
+        # raise ValueError("Callback functions must have a ‘__call__’ attribute.")
+        self._callbacks.append(func)
+
+
+class QMFileValidator(object):
+    """
+    Function: Sets up a series of functions that take in a
+    dataframe, check to see if it has any internal conflicts,
+    and throws an error if one is found.
+    """
+    def __init__(self, errorType):
+        self._callbacks = []
+        self._errorType = errorType
+
+    def __call__(self, filename):
+        """
+        Function: Cycles through each callback, feeding the
+        dataframe to the next function to check for errors.
+        :param df: dataframe to be tested
+        :param filename: Filename corresponding to the dataframe
+        :param sheet: Sheet corresponding to the dataframe
+        :return: dataframe with no errors
+        """
+        for callback in self._callbacks:
+            if callback(filename):
+                _logger.info(f"File '{filename}' exists.")
+                return
+        self.logger(self._errorType)(f"File '{filename}' has failed.")
+        raise ValueError()
+
+    def add_callback(self, func):
+        # if not hasattr('__call__', func):
+        #     raise ValueError("Callback functions must have a ‘__call__’ attribute.")
+        self._callbacks.append(func)
+
+    def logger(self, type):
+        """
+        Functionality: Returns the correct logger function
+        :param type: Error type to be thrown by the logger.
+        :return: Logger function that corresponds the the type.
+        """
+        return {
+            'DEBUG': _logger.debug,
+            'INFO': _logger.info,
+            'WARNING': _logger.warning,
+            'ERROR': _logger.error,
+            'CRITICAL': _logger.critical
+        }.get(type.upper(), 'ERROR')
+
+
+def validate_files(filenames, errorType='ERROR'):
+    validate = QMFileValidator(errorType)
+    validate.add_callback(check_file_path_exists)
+    validate.add_callback(check_url_exists)
+    e = None
+    for file in filenames:
+        try:
+            _logger.info(f"Validating {file}")
+            validate(file)
+        except ValueError:
+            e = True
+            continue
+            #  Only raises error at the end
+    if e:
+        _logger.info("One or more files failed validation")
```

Comment:
Then you can report on all the files that failed.

```python
msg = '\t\n'.join(['The following files failed validation'] + badFiles)
_logger.error(msg)
raise ValueError(msg)
```

I don't like the use of `ValueError` here, but I don't know that any other really fits. We should create our own exception. Define the exception at the top of this file. (Yes, it really is that simple!)

```python
class InvalidFileError(Exception):
    pass
```

Then the code above becomes

```python
msg = '\t\n'.join(['The following files failed validation'] + badFiles)
_logger.error(msg)
raise InvalidFileError(msg)
```